aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+
+  auto Bs = at::native::_compute_linear_combination(As, bs);
+
+  // compute A6
+  Bs.select(0, 2).add_(at::matmul(Bs.select(0, 3), Bs.select(0, 3)));
+  return Bs.select(0,0).add_(at::matmul(
+    Bs.select(0, 1).add_(Bs.select(0, 2)),
+    Bs.select(0, 2)
+  ));
+}
+
+template <typename scalar_t>
+Tensor compute_T18(const Tensor& A) {
+  constexpr int num_prods = 5;
+  array2d<scalar_t, num_prods, num_prods> b = {{
+    {
+      0.,
+      -1.00365581030144618291e-01,
+      -8.02924648241156932449e-03,
+      -8.92138498045729985177e-04,
+      0.
+    },
+    {
+      0.,
+      3.97849749499645077844e-01,
+      1.36783778460411720168e+00,
+      4.98289622525382669416e-01,
+      -6.37898194594723280150e-04
+    },
+    {
+      -1.09676396052962061844e+01,
+      1.68015813878906206114e+00,
+      5.71779846478865511061e-02,
+      -6.98210122488052056106e-03,
+      3.34975017086070470649e-05
+    },
+    {
+      -9.04316832390810593223e-02,
+      -6.76404519071381882256e-02,
+      6.75961301770459654925e-02,
+      2.95552570429315521194e-02,
+      -1.39180257516060693404e-05
+    },
+    {
+      0.,
+      0.,
+      -9.23364619367118555360e-02,
+      -1.69364939002081722752e-02,
+      -1.40086798182036094347e-05
+    }
+  }};
+
+  // gather coefficients `b` from above into a tensor,
+  // and move them to device `device_of(A)`
+  auto bs = at::from_blob(
+    reinterpret_cast<void*>(&b),
+    {num_prods, num_prods},
+    {num_prods, 1},
+    A.dtype()
+  );
+  bs = _pin_and_move_memory_if_cuda_input(bs, A);
+
+  auto As = _allocate_buffer(A, num_prods);
+  _fill_matrix_powers(As, A, num_prods);
+
+  auto Bs = at::native::_compute_linear_combination(As, bs);
+
+  // compute A9
+  Bs.select(0, 3).add_(at::matmul(Bs.select(0, 0), Bs.select(0, 4)));
+  return Bs.select(0, 1).add_(at::matmul(
+    Bs.select(0, 2).add_(Bs.select(0, 3)),
+    Bs.select(0, 3)
+  ));
+}
+
+template <typename scalar_t>
+void compute_T18_scale_square(
+  Tensor& mexp_out,
+  const Tensor& a,
+  const Tensor& norm,
+  scalar_t theta
+) {
+  // Scale
+  const auto s = at::max(
+    at::zeros_like(norm),
+    at::ceil(at::log2(norm / theta))
+  ).unsqueeze(-1).unsqueeze(-1).to(at::kLong);
+  const auto pow2s = at::pow(2, s);
+  const auto a_scaled = a / pow2s;
+
+  // Square
+  auto mexp_scaled = at::native::compute_T18<scalar_t>(a_scaled);
+  auto s_cpu = s.to(at::kCPU);
+  for (int64_t i = 0; i < mexp_scaled.size(0); ++i) {
+    auto s_val = s_cpu.select(0, i).template item<int64_t>();
+    auto mexp = mexp_scaled.select(0, i);
+    for (int64_t p = 0; p < s_val; ++p) {
+      mexp = at::matmul(mexp, mexp);
+    }
+    mexp_out.select(0, i).copy_(mexp);
+  }
+}
+
+template <typename scalar_t>
+Tensor mexp_impl(
+  const Tensor& a,
+  std::array<scalar_t, total_n_degs> thetas,
+  bool compute_highest_degree_approx = false
+) {
+  auto res = at::empty_like(a);
+  const auto norm = operator_1_norm(a);
+  const auto norm_cpu = norm.to(at::kCPU);
```

Comment:
Don't do this here!!! Do it in the conditional lol

Reply:
One question I'll be asking myself as I read the rest of this function is why you need an entire honking tensor to store this scalar, esp. when this function is already templated over `scalar_t`. It is probably lower overhead and more convenient to work with if you just get the norm into a `scalar_t` for later.

Reply:
Fixed.